h2o-core/src/main/java/water/parser/ParseSetup.java

```diff
@@ -267,11 +273,16 @@ public static ParseSetup guessSetup( Key[] fkeys, ParseSetup userSetup ) {
     //Calc chunk-size
     // FIXME: should be a parser specific - or at least parser should be able to override defaults
     Iced ice = DKV.getGet(fkeys[0]);
+    
     if (ice instanceof Frame && ((Frame) ice).vec(0) instanceof UploadFileVec) {
       t._gblSetup._chunk_size = FileVec.DFLT_CHUNK_SIZE;
     } else {
-      t._gblSetup._chunk_size = FileVec.calcOptimalChunkSize(t._totalParseSize, t._gblSetup._number_columns, t._maxLineLength,
-              Runtime.getRuntime().availableProcessors(), H2O.getCloudSize(), false /*use new heuristic*/, true);
+      int calculatedChunkSize = FileVec.calcOptimalChunkSize(t._totalParseSize, t._gblSetup._number_columns, t._maxLineLength,
+          Runtime.getRuntime().availableProcessors(), H2O.getCloudSize(), false /*use new heuristic*/, true);
+      int maxExpectedHeaderSize = Math.max(10000, (int)t._maxLineLength * 300); // to make sure we dont lose bits and get a negative
+      int acceptableHeaderSize = Math.min(500000, maxExpectedHeaderSize); //can't expect half a meg of headers
+      int finalChunkSize = Math.max(calculatedChunkSize, acceptableHeaderSize);
```

Comment:
Yes, it is 2017 - but we have a function that is supposed to calculate the optimal chunk size FileVec.calcOptimalChunkSize(...). This function contains the heuristics that determines the chunk size. If you are overriding the result of the function you are by _definition of the function_ creating a sub-optimal chunk size (assuming the calcOptimalChunkSize function really returns the optimal chunk size, that would be another discussion).

What part of the ARFF parser requires the header to fit in just one chunk? Is it the guessSetup function? Please note that the AbstractParserProvide provides a ByteVec and this lets you read as much data as you want in the guessSetup(...).